presto-main/src/main/java/com/facebook/presto/type/Re2JRegexp.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.type;
+
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.google.re2j.Matcher;
+import com.google.re2j.Options;
+import com.google.re2j.Pattern;
+import io.airlift.log.Logger;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.primitives.Ints.checkedCast;
+import static com.google.re2j.Options.Algorithm.DFA_FALLBACK_TO_NFA;
+import static java.lang.String.format;
+
+public final class Re2JRegexp
+{
+    private static final Logger log = Logger.get(Re2JRegexp.class);
+
+    private static final java.util.regex.Pattern DOT_STAR_PREFIX_SUFFIX_PATTERN = java.util.regex.Pattern.compile("^(\\.\\*\\??)?(.*?)(\\.\\*\\??)?$");
+    private static final int CORE_PATTERN_INDEX = 2;
+
+    public final int dfaStatesLimit;
+    public final int dfaRetries;
+
+    public final Pattern re2jPattern;
+    public final Pattern re2jPatternWithoutDotStartPrefixSuffix;
+
+    public Re2JRegexp(int dfaStatesLimit, int dfaRetries, Slice pattern)
+    {
+        this.dfaStatesLimit = dfaStatesLimit;
+        this.dfaRetries = dfaRetries;
+
+        Options options = Options.builder()
+                .setAlgorithm(DFA_FALLBACK_TO_NFA)
+                .setMaximumNumberOfDFAStates(dfaStatesLimit)
+                .setNumberOfDFARetries(dfaRetries)
+                .setEventsListener(new RE2JEventsListener())
+                .build();
+
+        String patternString = pattern.toStringUtf8();
+        re2jPattern = Pattern.compile(patternString, options);
+
+        // Remove .* prefix and suffix. This will give performance boost when using RE2J Pattern.find() function.
+        java.util.regex.Matcher dotStarPrefixSuffixMatcher = DOT_STAR_PREFIX_SUFFIX_PATTERN.matcher(patternString);
+        checkState(dotStarPrefixSuffixMatcher.matches());
+        String patternStringWithoutDotStartPrefixSuffix = dotStarPrefixSuffixMatcher.group(CORE_PATTERN_INDEX);
+
+        if (!patternStringWithoutDotStartPrefixSuffix.equals(patternString)) {
+            re2jPatternWithoutDotStartPrefixSuffix = Pattern.compile(patternStringWithoutDotStartPrefixSuffix, options);
+        }
+        else {
+            re2jPatternWithoutDotStartPrefixSuffix = re2jPattern;
+        }
+    }
+
+    public boolean matches(Slice source)
+    {
+        return re2jPatternWithoutDotStartPrefixSuffix.find(source);
```

Comment:
I had a discussion with @martint about this in person. He's also very concerned about this.

@sopel39, is the `.*` stripping intended to be temporary or permanent? Does this problem exist in RE2 for C++?

I'm posting the benchmark result here to help future discussions (on my MacBook Retina laptop):
```
RegexpFunctionsBenchmark.benchmarkLikeJoni            .*x.*            1024  avgt   10    882.045 ±   56.641  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni            .*x.*           32768  avgt   10  26236.741 ±  424.001  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni        .*(x|y).*            1024  avgt   10   1212.163 ±   74.070  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni        .*(x|y).*           32768  avgt   10  35566.473 ±  638.837  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni      longdotstar            1024  avgt   10    571.628 ±    9.082  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni      longdotstar           32768  avgt   10  15415.762 ±  387.839  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni            phone            1024  avgt   10   2505.099 ±   96.680  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni            phone           32768  avgt   10   8453.426 ±  135.765  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni          literal            1024  avgt   10    840.726 ±   20.169  ns/op
RegexpFunctionsBenchmark.benchmarkLikeJoni          literal           32768  avgt   10  25820.271 ±  356.881  ns/op

RegexpFunctionsBenchmark.benchmarkLikeRe2J            .*x.*            1024  avgt   10   1246.662 ±   41.722  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            .*x.*           32768  avgt   10  37210.596 ± 1932.559  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J        .*(x|y).*            1024  avgt   10   1265.604 ±   95.069  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J        .*(x|y).*           32768  avgt   10  36564.911 ± 2521.822  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J      longdotstar            1024  avgt   10   1307.066 ±  101.511  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J      longdotstar           32768  avgt   10  46224.595 ± 1233.231  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            phone            1024  avgt   10    238.278 ±    6.322  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            phone           32768  avgt   10    453.905 ±    5.114  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J          literal            1024  avgt   10   1333.154 ±   35.057  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J          literal           32768  avgt   10  49091.516 ± 1212.082  ns/op

// after removing .* stripping code
RegexpFunctionsBenchmark.benchmarkLikeRe2J            .*x.*            1024  avgt   10    5774.025 ±    113.456  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            .*x.*           32768  avgt   10  256218.996 ± 126455.868  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J        .*(x|y).*            1024  avgt   10    5885.465 ±    155.244  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J        .*(x|y).*           32768  avgt   10  228570.880 ± 108289.568  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J      longdotstar            1024  avgt   10    5747.038 ±    193.611  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J      longdotstar           32768  avgt   10  246181.409 ± 116813.297  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            phone            1024  avgt   10     566.565 ±     29.427  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J            phone           32768  avgt   10     129.442 ±      3.320  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J          literal            1024  avgt   10    1536.869 ±     52.667  ns/op
RegexpFunctionsBenchmark.benchmarkLikeRe2J          literal           32768  avgt   10   51700.115 ±   1027.582  ns/op
```